### PR TITLE
Issue 42355: Perf problem during copy-to-study with many rows

### PR DESCRIPTION
**Pull request by labkey-jeckels**
#### Rationale
Users want to be able to copy large assay runs to study datasets. It's painfully slow because we are iterating on each row and doing lots of DB queries

#### Changes
* Switch AssayProvider to encourage implementations to query rows in bulk
* Cache ExpData, ExpRun, and ExpExperiment references during process as they're likely to be shared by result rows
* Give user more feedback on what's happening during confirmation page
* Optimize date math for visit calculation

### CODE REVIEW

File: api/src/org/labkey/api/assay/AbstractAssayProvider.java [hunk 1]
```diff
@@ -757,52 +758,106 @@ public ExpProtocol createAssayDefinition(User user, Container container, String
...
         {
+            Container container = null;
+            if (data.getRunId() != null)
+            {
+                ExpRun run = cache.getRun(data.getRunId());
+                if (run != null)
+                {
+                    // Ignore Results domain TargetStudy for now.
+                    // The participant resolver will find the TargetStudy on the row.
+                    ExpObject source = switch (pair.first)
```
Comment (labkey-kevink) [inline api/src/org/labkey/api/assay/AbstractAssayProvider.java:783]: Ooh, switch expression is fun!

File: api/src/org/labkey/api/assay/AbstractAssayProvider.java [hunk 2]
```diff
@@ -757,52 +758,106 @@ public ExpProtocol createAssayDefinition(User user, Container container, String
...
+                            container = ContainerManager.getForId(targetStudyId);
+                    }
+                }
+            }
+            result.add(container);
         }
+        return result;
+    }
+
+//    @Override
```
Comment (labkey-kevink) [inline api/src/org/labkey/api/assay/AbstractAssayProvider.java]: remove?

File: api/src/org/labkey/api/assay/AbstractAssayProvider.java [hunk 3]
```diff
@@ -757,52 +758,106 @@ public ExpProtocol createAssayDefinition(User user, Container container, String
...
+
+        private <K, T extends ExpObject> T get(K key, Map<K, T> cache, Supplier<T> supplier)
         {
-            Map<String, Object> properties = OntologyManager.getProperties(source.getContainer(), source.getLSID());
-            String targetStudyId = (String) properties.get(targetStudyColumn.getPropertyURI());
+            if (key == null)
+            {
+                return null;
+            }
+            if (cache.containsKey(key))
```
Comment (labkey-kevink) [inline api/src/org/labkey/api/assay/AbstractAssayProvider.java:830]: Can this be replaced with `cache.computeIfAbsent(key, k -> supplier.get())`?

File: study/src/org/labkey/study/model/StudyImpl.java [hunk 4]
```diff
@@ -1233,7 +1234,19 @@ public static int daysBetween(Calendar startDate, Calendar endDate)
...
+        int daysBetween = (int)(diffInMillis / TimeUnit.DAYS.toMillis(1));
+
+        if (daysBetween > 0)
+        {
+            // Round down to ensure consistent results with original implementation, which added one day at a time and
+            // was a performance bottleneck when the supplied dates were many years apart
+            daysBetween--;
+            date.add(Calendar.DAY_OF_MONTH, daysBetween);
+        }
+
```
Comment (labkey-kevink) [inline study/src/org/labkey/study/model/StudyImpl.java:1249]: Shouldn't the while loop after this change be removed now that are calculating the date diff more efficiently?

**Review by labkey-kevink — APPROVED** [review]

File: api/src/org/labkey/api/assay/AbstractAssayProvider.java (hunk 2 above)
Comment (labkey-jeckels) [reply api/src/org/labkey/api/assay/AbstractAssayProvider.java]: Yes

File: api/src/org/labkey/api/assay/AbstractAssayProvider.java (hunk 3 above)
Comment (labkey-jeckels) [reply api/src/org/labkey/api/assay/AbstractAssayProvider.java:830]: computeIfAbsent() computes if the value is null, which means that it wouldn't cache misses. I'll add a comment.

File: study/src/org/labkey/study/model/StudyImpl.java (hunk 4 above)
Comment (labkey-jeckels) [reply study/src/org/labkey/study/model/StudyImpl.java:1249]: Just pushed a test case to ensure consistency with the original (and questionable) implementation, plus comments.

**Review by labkey-klum — APPROVED** [review]

**Pull request merged**